federation/cmd/federation-apiserver/app/server.go

```diff
@@ -221,8 +232,11 @@ func Run(s *options.ServerRunOptions) error {
 	installBatchAPIs(m, genericConfig.RESTOptionsGetter)
 	installAutoscalingAPIs(m, genericConfig.RESTOptionsGetter)
 
-	sharedInformers.Start(wait.NeverStop)
-	return m.PrepareRun().Run(wait.NeverStop)
+	err = m.PrepareRun().NonBlockingRun(stopCh)
+	if err == nil {
+		sharedInformers.Start(stopCh)
```

Comment:
The typical flow that "just works" is like this:

1. Set up anything that uses a shared informer, for example:
    1. `sharedInformers.Core().V1().Pods().AddEventHandler(...)`
1. If you're using a lister from a shared informer, make sure you wait for its informer to sync first, for example:
    1. `if ! cache.WaitForCacheSync(stopCh, podListerSynced, ...)`
1. Start the shared informer factory via `sharedInformers.Start(...)`

Step 1 is required. Step 2 is optional but highly encouraged.

If the first time you reference a shared informer is after you've started the factory, you'll have to manually run that shared informer (we should probably fix that) cc @deads2k.